### PR TITLE
Introduce message bus

### PR DESCRIPTION
**Pull request by sven1103**
This PR introduces a simple Notification service and messag bus, that can be used to broadcast events in a simple publish/subscribe pattern.

### CODE REVIEW

File: domain/src/main/java/life/qbic/apps/datamanager/notifications/MessageParameters.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+/**
+ * <b><class short description - 1 Line!></b>
+ *
+ * <p><More detailed description - When to use, what it solves, etc.></p>
+ *
+ * @since <version tag>
+ */
+public final class MessageParameters {
+
+  public final String messageType;
```
Comment (wow-such-code) [inline domain/src/main/java/life/qbic/apps/datamanager/notifications/MessageParameters.java:14]: why is this a string and not an enum?

File: domain/src/main/java/life/qbic/apps/datamanager/notifications/MessageParameters.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  private MessageParameters(String messageType, String messageId, String occuredOn) {
+    super();
+    this.messageId = messageId;
+    this.messageType = messageType;
+    this.occuredOn = occuredOn;
+  }
+
+  public static MessageParameters durableTextParameters(String messageType, String messageId,
+      Instant occuredOn) {
+    String occuredOnString = occuredOn.toString();
```
Comment (wow-such-code) [inline domain/src/main/java/life/qbic/apps/datamanager/notifications/MessageParameters.java:29]: is it bad practice to do this in the constructor? no critique, just curiousity

File: domain/src/main/java/life/qbic/apps/datamanager/notifications/MessageParameters.java (hunk 1 above)
Comment (wow-such-code) [reply domain/src/main/java/life/qbic/apps/datamanager/notifications/MessageParameters.java:14]: ah, because this is taken from class names?

Comment (sven1103) [reply domain/src/main/java/life/qbic/apps/datamanager/notifications/MessageParameters.java:14]: exactly

File: domain/src/main/java/life/qbic/apps/datamanager/notifications/MessageParameters.java (hunk 2 above)
Comment (sven1103) [reply domain/src/main/java/life/qbic/apps/datamanager/notifications/MessageParameters.java:29]: Basically what we learned from Joshua Bloch's Effective Java Item 1: Consider static factory methods over constructors. If you follow the newer Java API, you will find it everywhere. 

https://dev.to/kylec32/effective-java-tuesday-let-s-consider-static-factory-methods-170p

File: domain/src/main/java/life/qbic/apps/datamanager/notifications/NotificationService.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            message = ObjectSerializer.instance().serialise(notification.event);
+        } catch (IOException e) {
+            throw new ServiceException("Notification was not send", e.getCause());
+        }
+
+        messageBus.submit(message, messageParams);
+    }
+
+    /**
+     * Generates a unique notification id.
```
Comment (wow-such-code) [inline domain/src/main/java/life/qbic/apps/datamanager/notifications/NotificationService.java:53]: how unique is it? :)

File: domain/src/main/java/life/qbic/domain/events/DomainEventProducer.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
+package life.qbic.domain.events;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * <b><class short description - 1 Line!></b>
+ *
+ * <p><More detailed description - When to use, what it solves, etc.></p>
```
Comment (wow-such-code) [inline domain/src/main/java/life/qbic/domain/events/DomainEventProducer.java:9]: missing description

File: domain/src/main/java/life/qbic/domain/events/DomainEventProducer.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+ *
+ * @since <version tag>
+ */
+public class DomainEventProducer {
+
+  private static final ThreadLocal<List> subscribers = new ThreadLocal<>();
+
+  private static final ThreadLocal<Boolean> publishing = ThreadLocal.withInitial(
+      () -> Boolean.FALSE);
```
Comment (wow-such-code) [inline domain/src/main/java/life/qbic/domain/events/DomainEventProducer.java:18]: what does this do?

**Review by wow-such-code — CHANGES_REQUESTED** [review]
Looks very good to me, but I do have a few questions and one NIT (missing description in one class)

File: domain/src/main/java/life/qbic/domain/events/DomainEventProducer.java (hunk 5 above)
Comment (sven1103) [reply domain/src/main/java/life/qbic/domain/events/DomainEventProducer.java:18]: https://www.baeldung.com/java-threadlocal

We basically want to store the current state of the DomainEvent producer in the local thread (which is the user session). It is a flag, that prevents from subscribing when the DomainEventProducer is currently publishing.

File: domain/src/main/java/life/qbic/apps/datamanager/notifications/NotificationService.java (hunk 3 above)
Comment (sven1103) [reply domain/src/main/java/life/qbic/apps/datamanager/notifications/NotificationService.java:53]: https://en.wikipedia.org/wiki/Universally_unique_identifier#Version_4_(random)

unique enough :D

**Review by wow-such-code — APPROVED** [review]
nice

**Review by wow-such-code — APPROVED** [review]

**Review by wow-such-code — APPROVED** [review]

**Pull request merged**